control_panel_api/helm.py

```diff
@@ -0,0 +1,35 @@
+from django.conf import settings
+
+
+def _get_subprocess():
+    """Allow the subprocess module to be switched out at runtime for a
+    magicmock to prevent real calls during testing
+    """
+    return settings.SUBPROCESS_MODULE
```

Comment:
In retrospect, I think this approach to swapping out modules for test purposes is not so good.
1. Apart from the tests, the app has no need for this setting. This simplifies the test code at the expense of the app code.
2. It also moves the point where the substitution of the module happens away from the tests in which it is used. The tests should be explicit about their setup.

The tests should instead monkey-patch the `helm_upgrade` function, eg:
```python
@mock.patch('control_panel_api.models.helm.helm_upgrade')
```
or even `subprocess.run`
```python
@mock.patch('control_panel_api.helm.subprocess.run')
```

Reply:
I initially had done that with the aws stuff but required multiple patches on classes, functions etc and still was accidentally called.

How would you suggest a simple and effective way so don't have to litter code with patches and accidentally miss any?  It could get called from a view or model etc

We can patch individual tests, or at class level.  Is there a better way?

Reply:
Yeah, it is a tricky problem and I'm now thinking of a solution which is similar to what you already have.

Instead of having the modules to swap out as settings, maybe we need something like feature flags - settings like `ENABLE_FEATURE=True`, maybe a dict of them if there are lots, and then the code could wrap (in this case) the call to `subprocess.run` with an `if ENABLE_SUBPROCESS` or `if ENABLE_HELM` and set them all to `False` in the test settings.

Although it's basically the same thing as what you've got, it seems a bit more straightforward, and we should probably use feature flags for new features we're developing anyway.

Reply:
I think that would soften the first issue I had, because the code would be self-documenting.
I'm not sure if having a version of the settings for tests is so good, as per my second issue. It would be more explicit to override the settings near the test code.

Reply:
Yes agree it's a tricky one, I sort of see what you're saying.  

The test settings is a pattern I've seen and used before - commonly switching out things like cache and session middlewares with dummy implementations e.g. https://docs.djangoproject.com/en/1.11/topics/cache/#dummy-caching-for-development

There's also `@override_settings` could help.

I was using the magicmock as it just gives you the interface and works - now noticing that's just a lazy version of the empty dummy cache class.

A better design could be a simple command runner class or interface that uses subprocess, which then could be switched out at runtime using the test environment for a dummy runner?

Not sure, further research required!